src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs

```diff
@@ -0,0 +1,470 @@
+                    // manually advance the stream pointer to the next header before doing the substitution
+                    // The original stream will get disposed when the reader gets disposed.
+                    _readerOfOrigin.AdvanceDataStreamIfNeeded();
+                    // We only do this once
+                    _readerOfOrigin = null;
+                }
+
+                _header._dataStream?.Dispose();
+
+                _header._dataStream = value;
+            }
+        }
+
+        /// <summary>
+        /// A string that represents the current entry.
+        /// </summary>
+        /// <returns>The <see cref="Name"/> of the current entry.</returns>
+        public override string ToString() => Name;
+
+        // Abstract method that determines if setting the data stream for this entry is allowed.
+        internal abstract bool IsDataStreamSetterSupported();
+
+        // Extracts the current entry to a location relative to the specified directory.
+        internal void ExtractRelativeToDirectory(string destinationDirectoryPath, bool overwrite)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(destinationDirectoryPath));
+            Debug.Assert(Path.IsPathFullyQualified(destinationDirectoryPath));
+
+            string destinationDirectoryFullPath = destinationDirectoryPath.EndsWith(Path.DirectorySeparatorChar) ? destinationDirectoryPath : destinationDirectoryPath + Path.DirectorySeparatorChar;
+
+            string fileDestinationPath = GetSanitizedFullPath(destinationDirectoryFullPath, Name, SR.TarExtractingResultsFileOutside);
+
+            string? linkTargetPath = null;
+            if (EntryType is TarEntryType.SymbolicLink or TarEntryType.HardLink)
+            {
+                if (string.IsNullOrEmpty(LinkName))
+                {
+                    throw new FormatException(SR.TarEntryHardLinkOrSymlinkLinkNameEmpty);
+                }
+
+                linkTargetPath = GetSanitizedFullPath(destinationDirectoryFullPath, LinkName, SR.TarExtractingResultsLinkOutside);
+            }
+
+            if (EntryType == TarEntryType.Directory)
+            {
+                Directory.CreateDirectory(fileDestinationPath);
+            }
+            else
+            {
+                // If it is a file, create containing directory.
+                Directory.CreateDirectory(Path.GetDirectoryName(fileDestinationPath)!);
+                ExtractToFileInternal(fileDestinationPath, linkTargetPath, overwrite);
+            }
+
+            // If the path can be extracted in the specified destination directory, returns the full path with sanitized file name. Otherwise, throws.
+            static string GetSanitizedFullPath(string destinationDirectoryFullPath, string path, string exceptionMessage)
+            {
+                string actualPath = Path.Join(Path.GetDirectoryName(path), ArchivingUtils.SanitizeEntryFilePath(Path.GetFileName(path)));
```

Comment:
perf: since `Path.GetFileName` returns only a substring of the provided path, we could use the overload that returns `ReadOnlySpan<char>` here and make `SanitizeEntryFilePath` accept `ROS` and avoid additional string allocation